File: server/src/main/java/org/queue/consumer/ConsumerConnector.scala
/*
 * Copyright 2010 LinkedIn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.queue.consumer

import org.queue.utils.Utils
import org.slf4j.LoggerFactory
import org.slf4j.event.Level

/**
 *  Main interface for consumer
 */
trait ConsumerConnector {

  def createMessageStreams(topicCountMap: Map[String,Int]) : Map[String,List[KafkaMessageStream]]

  /**
   *  Commit the offsets of all broker partitions connected by this connector.
   */
  def commitOffsets()
  
  /**
   *  Shut down the connector
   */
  def shutdown()
}

object Consumer {
  private val logger = LoggerFactory.getLogger(getClass())
  private val consumerStatsMBeanName = "queue:type=queue.ConsumerStats"

  /**
   *  Create a ConsumerConnector
   *
   *  @param config  at the minimum, need to specify the groupid of the consumer and the zookeeper
   *                 connection string zk.connect.
   */
  def create(config: ConsumerConfig): ConsumerConnector = {
    val consumerConnect = new ZookeeperConsumerConnector(config)
    Utils.swallow(Level.TRACE,Utils.registerMBean(consumerConnect, consumerStatsMBeanName))
    consumerConnect
  }

  /**
   *  Create a ConsumerConnector
   *
   *  @param config  at the minimum, need to specify the groupid of the consumer and the zookeeper
   *                 connection string zk.connect.
   */
  def createJavaConsumerConnector(config: ConsumerConfig): org.queue.javaapi.consumer.ConsumerConnector = {
    val consumerConnect = new org.queue.javaapi.consumer.ZookeeperConsumerConnector(config)
    Utils.swallow(Level.WARN, Utils.registerMBean(consumerConnect.underlying, consumerStatsMBeanName))
    consumerConnect
  }
}

File: server/src/main/java/org/queue/consumer/ConsumerConnector.java
package org.queue.consumer;

import java.util.Map;

/**
 * @author gaoyvfeng
 * @ClassName ConsumerConnector
 * @description:
 * @datetime 2024年 05月 22日 17:39
 * @version: 1.0
 */
public interface ConsumerConnector {
    /**
     * 创建消息流。
     * @param topicCountMap 主题到分区数的映射。
     * @return 按主题分组的Kafka消息流列表。
     */
    Map<String, List<KafkaMessageStream>> createMessageStreams(Map<String, Integer> topicCountMap);

    /**
     * 提交此连接器连接的所有代理分区的偏移量。
     */
    void commitOffsets();

    /**
     * 关闭连接器。
     */
    void shutdown();
}
